docs/spec/staking/end_block.md

```diff
@@ -1,68 +1,40 @@
-        } else {
-            validator.Status = sdk.Unbonded
-            SetValidator(unbondingValidator)
-        }
-    return
-```
+ - The new validator set is taken as the top `params.MaxValidators` number of
+   validators retrieved from the ValidatorsByPower index
+ - The previous validator set is compared with the new validator set 
+   - missing validators begin unbonding
+   - new validator are instantly bonded
 
-## Validator Set Changes
+in all cases any validators leaving or entering the bonded validator set or
+changing balances and staying within the bonded validator set incure an update
+message which is passed back to Tendermint.
 
-The Tendermint validator set may be updated by state transitions that run at
-the end of every block. The Tendermint validator set may be changed by
-validators either being jailed due to inactivity/unexpected behaviour (covered
-in slashing) or changed in validator power. Determining which validator set
-changes must be made occurs during staking transactions (and slashing
-transactions) - during end-block the already accounted changes are applied and
-the changes cleared
+## Queues 
 
-```golang
-EndBlock() ValidatorSetChanges
-    vsc = GetValidTendermintUpdates()
-    ClearTendermintUpdates()
-    return vsc
-```
+### Unbonding Validators
 
-## CompleteUnbonding
+Each block the validator queue is to be checked for mature unbonding
+validators.  For all unbonding validators that have finished their unbonding
+period, the validator.Status is switched from sdk.Unbonding to sdk.Unbonded if
+they still have any delegation left.  Otherwise, the validator object is
+deleted from state.
 
-Complete the unbonding and transfer the coins to the delegate. Realize any
-slashing that occurred during the unbonding period.
+### Unbonding Delegations
 
-```golang
-unbondingQueue(currTime time.Time):
-    // unbondings are in ordered queue from oldest to newest
-    for all unbondings whose CompleteTime < currTime:
-        validator = GetValidator(unbonding.ValidatorAddr)
-        AddCoins(unbonding.DelegatorAddr, unbonding.Balance)
-        removeUnbondingDelegation(unbonding)
-    return
-```
+Complete the unbonding of all mature `UnbondingDelegations.Entries` by
+transferring the balance coins to the delegator's wallet address and removing
+the mature `UnbondingDelegation.Entries`. If there are no remaining entries also
+remove the `UnbondingDelegation` object from the store. 
 
-## CompleteRedelegation
+### Redelegations
 
-Note that unlike CompleteUnbonding slashing of redelegating shares does not
-take place during completion. Slashing on redelegated shares takes place
-actively as a slashing occurs. The redelegation completion queue serves simply to
-clean up state, as redelegations older than an unbonding period need not be kept,
-as that is the max time that their old validator's evidence can be used to slash them.
+Complete the unbonding of all mature `Redelegation.Entries` by removing the
```

Comment:
ditto

Reply:
same comment as previously on state, updated the paragraph as well